src/tir/schedule/primitive/cache_read_write.cc

```diff
@@ -0,0 +1,792 @@
+    cache_region = region.value_or(BufferRegion::FullRegion(read_buffer));
+  } else {
+    // Cond 2. The buffer is written inside the block.
+    const StmtSRef& write_block_sref = _write_block_sref.value();
+    const BlockNode* write_block = TVM_SREF_TO_BLOCK(write_block, write_block_sref);
+    // Find the producing region
+    Optional<BufferRegion> region = RelatedBufferRegion(write_block->writes, read_buffer);
+    ICHECK(region.defined());
+    StmtSRef parent_sref = GetRef<StmtSRef>(write_block_sref->parent);
+
+    // Detect insert position
+    CacheLocDetector::Detect(self, write_block_sref, scope_sref, &info);
+    cache_region =
+        RelaxBufferRegion(self, region.value(), write_block_sref, parent_sref, info.loc_sref);
+  }
+
+  // Step 6. Making new cache stage block and rewrite readers.
+  Block cache_read_stage = MakeCacheStage(/*cache_region=*/cache_region, /*info=*/&info,
+                                          /*storage_scope=*/storage_scope);
+  Stmt new_scope = CacheReadRewriter::Rewrite(/*scope_sref=*/scope_sref, /*info=*/&info);
+
+  // Step 7. Replacing and updating flags.
+  self->Replace(scope_sref, new_scope, info.block_map);
+  StmtSRef result_block_sref = self->stmt2ref.at(cache_read_stage.get());
+  self->UpdateAffineFlag(result_block_sref);
+  BlockInfo& block_info = self->block_info[result_block_sref];
+  block_info.region_cover = true;
+  block_info.scope->stage_pipeline = true;
+  return result_block_sref;
+}
+
+StmtSRef CacheWrite(ScheduleState self, const StmtSRef& block_sref, int write_buffer_index,
+                    const String& storage_scope) {
+  /*!
+   * Check:
+   *   - The index is in the array of block reading region
+   *   - There is only one block who write the buffer in the scope
+   *
+   * Mutate:
+   *   - Allocate new cache buffer under the current scope.
+   *   - Find the lowest ancestor of the block and ANY ONE of the producer blocks.
+   *   - Copy the buffer with the consumed region.
+   */
+  // Step 1. Checking index
+  const BlockNode* block = TVM_SREF_TO_BLOCK(block, block_sref);
+  Buffer write_buffer =
+      GetNthAccessBuffer(self, GetRef<Block>(block), write_buffer_index, /*is_write=*/true);
+
+  // Step 2. Creating CacheStageInfo
+  CacheStageInfo info;
+  info.read_buffer = WithScope(write_buffer, storage_scope);
+  // Create the corresponding buffer to be written, i.e. result of cache_write
+  info.write_buffer = write_buffer;
+  // Create the corresponding buffer allocation
+  info.alloc = info.read_buffer;
+
+  // Step 3. Find the parent scope
+  StmtSRef scope_sref = GetScopeRoot(self, block_sref, /*require_stage_pipeline=*/true);
+  const BlockNode* scope_block = TVM_SREF_TO_BLOCK(scope_block, scope_sref);
+
+  // Step 4. Check the only writer block.
+  Optional<StmtSRef> _write_block_sref = GetOnlyWriteBlock(self, scope_sref, write_buffer);
+  // We have provide a block_sref who write the buffer, so use ICHECK here.
+  ICHECK(_write_block_sref.defined());
```

Comment:
Just feel free to unwrap

```suggestion
  ICHECK_EQ(block_sref.get(), GetOnlyWriteBlock(self, scope_sref, write_buffer).get());
```